consensus/core/src/dag_state.rs

```diff
@@ -176,19 +188,61 @@ impl DagState {
             unscored_committed_subdags,
             store,
             cached_rounds,
+            evicted_rounds: vec![0; num_authorities],
         };
 
         for (i, round) in last_committed_rounds.into_iter().enumerate() {
             let authority_index = state.context.committee.to_authority_index(i).unwrap();
-            let blocks = state
-                .store
-                .scan_blocks_by_author(
-                    authority_index,
-                    Self::eviction_round(round, cached_rounds) + 1,
-                )
-                .unwrap();
-            for block in blocks {
-                state.update_block_metadata(&block);
+            let blocks = if state.gc_enabled() {
+                // Find the latest block for the authority to calculate the eviction round. Then we want to scan and load the blocks from the eviction round and onwards only.
+                // As reminder, the eviction round is taking into account the gc_round.
+                let last_block = state
+                    .store
+                    .scan_last_blocks_by_author(authority_index, 1, None)
+                    .expect("Database error");
+                let last_block_round = last_block
+                    .last()
+                    .map(|b| b.round())
+                    .unwrap_or(GENESIS_ROUND);
+
+                let gc_eviction_round = Self::gc_eviction_round(
+                    last_block_round,
+                    state.gc_round(),
+                    state.cached_rounds,
+                ) + 1;
+
+                state
+                    .store
+                    .scan_blocks_by_author(authority_index, gc_eviction_round)
+                    .expect("Database error")
+            } else {
+                state
+                    .store
+                    .scan_blocks_by_author(
+                        authority_index,
+                        Self::eviction_round(round, cached_rounds) + 1,
+                    )
+                    .expect("Database error")
+            };
+
+            // Update the block metadata for the authority.
+            for block in &blocks {
+                state.update_block_metadata(block);
+            }
+
+            info!(
+                "Recoved blocks {}: {:?}",
+                authority_index,
+                blocks
+                    .iter()
+                    .map(|b| b.reference())
+                    .collect::<Vec<BlockRef>>()
+            );
+
+            // Update the evicted rounds for the authority now that all the blocks have been restored.
+            if state.gc_enabled() {
```

Comment:
Btw why this can't be updated earlier?

Reply:
I guess that's me getting a bit paranoid with the gc enabled checks. Yes it can.